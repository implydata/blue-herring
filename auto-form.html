<!--
IMPORTANT
This is a generated file. Don't modify anything in it manually. Use bork instead.
Bork (and bark) documentation can be found here : https://github.com/implydata/bark.
 -->
<!DOCTYPE HTML>
<html lang="en">
<head>
  <title>blue-herring | auto-form</title>

  <style type="text/css">a,abbr,acronym,address,applet,article,aside,audio,b,big,blockquote,body,button,canvas,caption,center,cite,code,dd,del,details,dfn,div,dl,dt,em,embed,fieldset,figcaption,figure,footer,form,h1,h2,h3,h4,h5,h6,header,html,i,iframe,img,ins,kbd,label,legend,li,mark,menu,nav,object,ol,output,p,pre,q,ruby,s,samp,section,small,span,strike,strong,sub,summary,sup,table,tbody,td,tfoot,th,thead,time,tr,tt,u,ul,var,video{margin:0;padding:0;border:0;font-size:100%;font:inherit;outline:none}article,aside,details,figcaption,figure,footer,header,menu,nav,section{display:block}body{line-height:1}ol,ul{list-style:none}table{border-collapse:collapse;border-spacing:0}input,textarea{outline:none;font-size:inherit;font-family:inherit}*{box-sizing:border-box;cursor:inherit}body *{position:relative;vertical-align:top}a{text-decoration:inherit;color:inherit;cursor:inherit}div,span{cursor:inherit}button,input{border:0;margin:0}text{cursor:default}svg *{fill:inherit}
</style>

  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.9.0/styles/solarized-light.min.css">
  <link href="https://fonts.googleapis.com/css?family=Open+Sans" rel="stylesheet">

  
    <link href="blueprint.css" rel="stylesheet">
  

  
    <link rel="stylesheet" type="text/css" media="screen" href="blue-herring.css">
  

  <link rel="stylesheet" type="text/css" media="screen" href="demo.css">

  <style type="text/css">
    
  </style>


</head>

<body>
  <div class="bork-header-bar">
    <div class="project-title">blue-herring's auto-form</div>
  </div>

  <div class="side-bar">
    
      
          <div class="side-bar-label">
            </a>
          </div>
      
    
      
          <a href="auto-form.html" class="side-bar-item selected">
            Auto Form</a>
          </a>
      
    
  </div>

  <div class="content">
    
      <div class="section demo ">
        <div class="view auto-form-demo-target"></div>
        <div class="section-label">Demo</div>
      </div>
    
      <div class="section code ">
        <pre><code class="jsx">import * as React from &#39;react&#39;;
import * as ReactDOM from &#39;react-dom&#39;;

import { AutoForm, AutoFormField } from &#39;blue-herring&#39;;

let schema: AutoFormField[] = [
  {key: &#39;dataSource&#39;, type: &#39;string&#39;},
  {key: &#39;parser&#39;, type: &#39;object&#39;, types: [
    {key: &#39;type&#39;, type: &#39;string&#39;},
    {key: &#39;parseSpec&#39;, type: &#39;object&#39;, types: [
      {key: &#39;format&#39;, type: &#39;string&#39;},
      {key: &#39;timestampSpec&#39;, type: &#39;object&#39;, types: [
        {key: &#39;column&#39;, type: &#39;string&#39;},
        {key: &#39;format&#39;, type: &#39;string&#39;}
      ]},
      {key: &#39;dimensionsSpec&#39;, type: &#39;object&#39;, types: [
        {key: &#39;dimensions&#39;, type: &#39;array&#39;, itemsTypes: [
          &#39;string&#39;,
          [{key: &#39;type&#39;, type: &#39;string&#39;}, {key: &#39;name&#39;, type: &#39;string&#39;}]
        ]},
        {key: &#39;dimensionExclusions&#39;, type: &#39;array&#39;, itemsTypes: [&#39;string&#39;]},
        {key: &#39;spatialDimensions&#39;, type: &#39;array&#39;, itemsTypes: [&#39;string&#39;]}
      ]},
    ]},
  ]},
  {key: &#39;metricsSpec&#39;, type: &#39;array&#39;, itemsTypes: [
    [{key: &#39;type&#39;, type: &#39;string&#39;}, {key: &#39;name&#39;, type: &#39;string&#39;}, {key: &#39;fieldName&#39;, type: &#39;string&#39;, optional: true}],
  ]},
  {key: &#39;granularitySpec&#39;, type: &#39;object&#39;, types: [
    {key: &#39;segmentGranularity&#39;, type: &#39;string&#39;},
    {key: &#39;queryGranularity&#39;, type: &#39;string&#39;},
    {
      key: &#39;intervals&#39;,
      type: &#39;date-interval&#39;,
      stringToInterval: (s: string) =&gt; s.split(&#39;/&#39;).map(_s =&gt; new Date(_s)) as [Date, Date],
      intervalToString: (dates: [Date, Date]) =&gt; dates[0].toLocaleDateString(&#39;en-US&#39;) + &#39;/&#39; + dates[1].toLocaleDateString(&#39;en-US&#39;)
    }
  ]}
];

let model: any = {
  dataSource: &quot;wikipedia&quot;,
  parser: {
    type: &quot;string&quot;,
    parseSpec: {
      format: &quot;json&quot;,
      timestampSpec: {
        column: &quot;timestamp&quot;,
        format: &quot;auto&quot;
      },
      dimensionsSpec: {
        dimensions: [
          &quot;page&quot;,
          &quot;language&quot;,
          &quot;user&quot;,
          &quot;unpatrolled&quot;,
          &quot;newPage&quot;,
          &quot;robot&quot;,
          &quot;anonymous&quot;,
          &quot;namespace&quot;,
          &quot;continent&quot;,
          &quot;country&quot;,
          &quot;region&quot;,
          &quot;city&quot;,
          {type: &quot;long&quot;, name: &quot;countryNum&quot;},
          {type: &quot;float&quot;, name: &quot;userLatitude&quot;},
          {type: &quot;float&quot;, name: &quot;userLongitude&quot;}
        ],
        dimensionExclusions: [],
        spatialDimensions: []
      }
    }
  },
  metricsSpec: [
    {type: &quot;count&quot;, name: &quot;count&quot;},
    {type: &quot;doubleSum&quot;, name: &quot;added&quot;, fieldName: &quot;added&quot;},
    {type: &quot;doubleSum&quot;, name: &quot;deleted&quot;, fieldName: &quot;deleted&quot;},
    {type: &quot;doubleSum&quot;, name: &quot;delta&quot;, fieldName: &quot;delta&quot;}
  ],
  granularitySpec: {
    segmentGranularity: &quot;DAY&quot;,
    queryGranularity: &quot;NONE&quot;,
    intervals: [&quot;2013-08-31/2013-09-01&quot;]
  },
  transformSpec: null
};

const onChange = (newValue: any) =&gt; {
  model = newValue;
  console.log(model)
  render();
};

const onSchemaChange = (newSchema: AutoFormField[]) =&gt; {
  schema = newSchema;
  render();
};

const render = () =&gt; {
  ReactDOM.render(

    &lt;AutoForm
      model={model}
      schema={schema}
      onChange={onChange}
      onSchemaChange={onSchemaChange}
    /&gt;

    ,
    document.getElementsByClassName(&#39;auto-form-demo-target&#39;)[0]
  );
};

render();
</code></pre>
        <div class="section-label">Demo code</div>
      </div>
    
  </div>

  <script crossorigin src="react.development.js"></script>
  <script crossorigin src="react-dom.development.js"></script>

  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.9.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script>

  <script src="vendor.js"></script>
  <script src="auto-form.js"></script>

</body>
</html>
