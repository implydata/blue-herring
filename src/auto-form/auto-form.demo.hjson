#
# Copyright (c) 2019 Imply Data, Inc. All rights reserved.
#
# This software is the confidential and proprietary information
# of Imply Data, Inc.

{
  tsx:
    '''
    # bork-hide
    import * as React from 'react';
    import * as ReactDOM from 'react-dom';
    # bork-hide-end

    import { AutoForm, AutoFormField } from 'blue-herring';

    const fields: AutoFormField[] = [
      {
        label: 'Foo',
        name: 'foo',
        type: 'string'
      }
    ];

    let model: any = {
      foo: 'Hello yes this is dog'
    };

    const onChange = (newValue: any) => {
      model = newValue;
      # bork-hide
      render();
      # bork-hide-end
    }

    # bork-hide
    const render = () => {
      ReactDOM.render(
    # bork-hide-end

    <AutoForm
      model={model}
      fields={fields}
      onChange={onChange}
    />

    # bork-hide
        ,
        document.getElementsByClassName('auto-form-demo-target')[0]
      );
    };

    render();
    # bork-hide-end

    '''
}
