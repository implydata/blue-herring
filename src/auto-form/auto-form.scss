.auto-form {
  > .auto-form {
    margin-left: 20px;
  }
}
