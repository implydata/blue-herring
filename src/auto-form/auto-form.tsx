import * as React from 'react';
import { AutoFormField, ObjectType, ArrayType, ArrayItemsType, PrimitiveType } from '../types/types';
import { Input } from '../input/input';

export interface AutoFormProps extends React.Props<any> {
  schema: AutoFormField[];
  model: any,
  onChange: (newValue: any) => void
  onSchemaChange?: (newSchema: AutoFormField[]) => void;
}

export class AutoForm extends React.Component<AutoFormProps, {}> {
  renderField = (field: AutoFormField) => {
    const { model, onChange } = this.props;

    return <Input
      key={field.key}
      field={field}
      model={model}
      onChange={v => onChange(Object.assign({}, model, {[field.key]: v}))}
    />;
  }

  render() {
    const { schema, model } = this.props;

    return <div className="auto-form">
      {model && schema.map(this.renderField)}
    </div>
  }
}
