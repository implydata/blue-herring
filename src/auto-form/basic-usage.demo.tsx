import * as React from 'react';
import * as ReactDOM from 'react-dom';

import { Example, Gallery } from 'hear-ye';

import { AutoForm } from './auto-form';
import { AutoFormField } from '../types/types';

export interface DemoState {
  schema?: AutoFormField[];
  model?: any;
}

class Demo extends React.Component<{}, DemoState> {
  constructor(props: {}) {
    super(props);

    this.state = {
      schema: [
        {key: 'dataSource', type: 'string'},
        {key: 'parser', type: 'object', types: [
          {key: 'type', type: 'string'},
          {key: 'parseSpec', type: 'object', types: [
            {key: 'format', type: 'string'},
            {key: 'timestampSpec', type: 'object', types: [
              {key: 'column', type: 'string'},
              {key: 'format', type: 'string'}
            ]},
            {key: 'dimensionsSpec', type: 'object', types: [
              {key: 'dimensions', type: 'array', itemsTypes: [
                'string',
                [{key: 'type', type: 'string'}, {key: 'name', type: 'string'}]
              ]},
              {key: 'dimensionExclusions', type: 'array', itemsTypes: ['string']},
              {key: 'spatialDimensions', type: 'array', itemsTypes: ['string']}
            ]},
          ]},
        ]},
        {key: 'metricsSpec', type: 'array', itemsTypes: [
          [{key: 'type', type: 'string'}, {key: 'name', type: 'string'}],
          [{key: 'type', type: 'string'}, {key: 'name', type: 'string'}, {key: 'fieldName', type: 'string'}],
        ]},
        {key: 'granularitySpec', type: 'object', types: [
          {key: 'segmentGranularity', type: 'string'},
          {key: 'queryGranularity', type: 'string'},
          {
            key: 'intervals',
            type: 'date-interval',
            stringToInterval: (s: string) => s.split('/').map(_s => new Date(_s)) as [Date, Date],
            intervalToString: (dates: [Date, Date]) => dates[0].toLocaleDateString('en-US') + '/' + dates[1].toLocaleDateString('en-US')
          }
        ]}
      ],
      model: {
        dataSource: "wikipedia",
        parser: {
          type: "string",
          parseSpec: {
            format: "json",
            timestampSpec: {
              column: "timestamp",
              format: "auto"
            },
            dimensionsSpec: {
              dimensions: [
                "page",
                "language",
                "user",
                "unpatrolled",
                "newPage",
                "robot",
                "anonymous",
                "namespace",
                "continent",
                "country",
                "region",
                "city",
                {type: "long", name: "countryNum"},
                {type: "float", name: "userLatitude"},
                {type: "float", name: "userLongitude"}
              ],
              dimensionExclusions: [],
              spatialDimensions: []
            }
          }
        },
        metricsSpec: [
          {type: "count", name: "count"},
          {type: "doubleSum", name: "added", fieldName: "added"},
          {type: "doubleSum", name: "deleted", fieldName: "deleted"},
          {type: "doubleSum", name: "delta", fieldName: "delta"}
        ],
        granularitySpec: {
          segmentGranularity: "DAY",
          queryGranularity: "NONE",
          intervals: ["2013-08-31/2013-09-01"]
        },
        transformSpec: null
      }
    }
  }

  onModelChange = (newModel: any) => {
    this.setState({model: newModel});
  }

  onSchemaChange = (newSchema: AutoFormField[]) => {
    this.setState({schema: newSchema});
  }

  render() {
    const { model, schema } = this.state;

    return <AutoForm
      model={model}
      schema={schema}
      onChange={this.onModelChange}
      onSchemaChange={this.onSchemaChange}
    />;
  }
}

Gallery.add({
  path: ['AutoForm'],
  component: <Demo/>
});
