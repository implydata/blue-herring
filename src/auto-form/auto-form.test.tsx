import * as React from 'react';

import { render, fireEvent } from 'react-testing-library';
import userEvent from 'user-event';
import { shallow, mount } from 'enzyme';

import { AutoForm, AutoFormField } from './auto-form';

describe('<AutoForm/>', () => {

  // react-testing-library
  it('fills a flat form at startup (react-testing-library)', () => {
    const fields: AutoFormField[] = [
      {name: "string", type: "string"},
      {name: "number", type: "number"},
      {name: "boolean", type: "boolean"},
      {name: "string_array", type: "string-array"},
      {name: "size_bytes", type: "size-bytes"}
    ];

    const originalModel = {
      string: 'Hello',
      number: 23,
      boolean: true,
      string_array: ['3', 'yes', 'this is dog'],
      size_bytes: 42
    };

    let model: typeof originalModel;

    const onChange = (newModel: typeof originalModel) => model = newModel;

    const comp = <AutoForm fields={fields} model={originalModel} onChange={onChange} />;


    const { container, getByValue, getByText } = render(comp);

    expect(container.firstChild).toMatchSnapshot();

    userEvent.type(getByValue('Hello'), '¡Hola!');
    expect(originalModel.string).toEqual('Hello');
    expect(model.string).toEqual('¡Hola!');

    userEvent.type(getByValue('23'), '42');
    expect(originalModel.number).toEqual(23);
    expect(model.number).toEqual(42);

    fireEvent.change(getByText('True').parentElement, {target: {value: 'False'}})
    expect(originalModel.boolean).toEqual(true);
    expect(model.boolean).toEqual(false);

    // userEvent.click(getByText('yes'));
    // userEvent.type(window, 'mwar');
    // fireEvent.keyPress(window, {which: 'enter'})
    // expect(originalModel.string_array).toEqual(['3', 'yes', 'this is dog']);
    // expect(model.string_array).toEqual(['3', 'mwar', 'this is dog'])

  });

  // Enzyme
  it('fills a flat form at startup (enzyme)', () => {
    const fields: AutoFormField[] = [
      {name: "string", type: "string"},
      {name: "number", type: "number"},
      {name: "boolean", type: "boolean"},
      {name: "string_array", type: "string-array"},
      {name: "size_bytes", type: "size-bytes"}
    ];

    const originalModel = {
      string: 'Hello',
      number: 23,
      boolean: true,
      string_array: ['3', 'yes', 'this is dog'],
      size_bytes: 42
    };

    let model: typeof originalModel;

    const onChange = (newModel: typeof originalModel) => model = newModel;

    const comp = <AutoForm fields={fields} model={originalModel} onChange={onChange} />;
    const shallowed = shallow(comp);
    expect(shallowed).toMatchSnapshot();

    const mounted = mount(comp);
    mounted.find('input[value="Hello"]').simulate('change', {target: {value: '¡Hola!'}});
    expect(originalModel.string).toEqual('Hello');
    expect(model.string).toEqual('¡Hola!');
  });
});
