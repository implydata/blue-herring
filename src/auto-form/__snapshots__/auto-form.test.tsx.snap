// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoForm/> fills a flat form at startup (enzyme) 1`] = `
<div
  className="auto-form"
>
  <Input
    field={
      Object {
        "key": "__root__",
        "label": "My awesome form",
        "type": "object",
        "types": Array [
          Object {
            "key": "string",
            "type": "string",
          },
          Object {
            "key": "number",
            "type": "number",
          },
          Object {
            "key": "boolean",
            "type": "boolean",
          },
          Object {
            "key": "size_bytes",
            "type": "size-bytes",
          },
        ],
      }
    }
    model={
      Object {
        "__root__": Object {
          "boolean": true,
          "number": 23,
          "size_bytes": 42,
          "string": "Hello",
          "string_array": Array [
            "3",
            "yes",
            "this is dog",
          ],
        },
      }
    }
    onChange={[Function]}
  />
</div>
`;

exports[`<AutoForm/> fills a flat form at startup (react-testing-library) 1`] = `
<div
  class="auto-form"
>
  <div
    class="object-input"
  >
    <div
      class="type-picker"
    >
      <button
        class="bp3-button bp3-minimal bp3-small folder"
        type="button"
      >
        <span
          class="bp3-icon bp3-icon-caret-down"
          icon="caret-down"
        >
          <svg
            data-icon="caret-down"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              caret-down
            </desc>
            <path
              d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 0 0-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
      <div
        class="type-picker-label"
      >
        My awesome form
      </div>
    </div>
    <div
      class="input object-field"
    >
      <div
        class="type-picker"
      >
        <div
          class="type-picker-label"
        >
          String
        </div>
      </div>
      <div
        class="bp3-input-group primitive-input"
      >
        <input
          class="bp3-input"
          style="padding-right: 10px;"
          type="text"
          value="Hello"
        />
      </div>
    </div>
    <div
      class="input object-field"
    >
      <div
        class="type-picker"
      >
        <div
          class="type-picker-label"
        >
          Number
        </div>
      </div>
      <div
        class="bp3-control-group bp3-numeric-input primitive-input"
      >
        <div
          class="bp3-input-group"
        >
          <input
            autocomplete="off"
            class="bp3-input"
            style="padding-right: 10px;"
            type="text"
            value="23"
          />
        </div>
        <div
          class="bp3-button-group bp3-vertical bp3-fixed"
        >
          <button
            class="bp3-button"
            type="button"
          >
            <span
              class="bp3-icon bp3-icon-chevron-up"
              icon="chevron-up"
            >
              <svg
                data-icon="chevron-up"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  chevron-up
                </desc>
                <path
                  d="M12.71 9.29l-4-4C8.53 5.11 8.28 5 8 5s-.53.11-.71.29l-4 4a1.003 1.003 0 0 0 1.42 1.42L8 7.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 0 0 .71-1.71z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
          <button
            class="bp3-button"
            type="button"
          >
            <span
              class="bp3-icon bp3-icon-chevron-down"
              icon="chevron-down"
            >
              <svg
                data-icon="chevron-down"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  chevron-down
                </desc>
                <path
                  d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 0 0-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0 0 12 5z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="input object-field"
    >
      <div
        class="type-picker"
      >
        <div
          class="type-picker-label"
        >
          Boolean
        </div>
      </div>
      <div
        class="bp3-html-select primitive-input"
      >
        <select>
          <option
            value="True"
          >
            True
          </option>
          <option
            value="False"
          >
            False
          </option>
        </select>
        <span
          class="bp3-icon bp3-icon-double-caret-vertical"
          icon="double-caret-vertical"
        >
          <svg
            data-icon="double-caret-vertical"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              double-caret-vertical
            </desc>
            <path
              d="M5 7h6a1.003 1.003 0 0 0 .71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 0 0 5 7zm6 2H5a1.003 1.003 0 0 0-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0 0 11 9z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="input object-field"
    >
      <div
        class="type-picker"
      >
        <div
          class="type-picker-label"
        >
          Size_bytes
        </div>
      </div>
      <div
        class="bp3-control-group bp3-numeric-input primitive-input"
      >
        <div
          class="bp3-input-group"
        >
          <input
            autocomplete="off"
            class="bp3-input"
            min="0"
            style="padding-right: 10px;"
            type="text"
            value="42"
          />
        </div>
        <div
          class="bp3-button-group bp3-vertical bp3-fixed"
        >
          <button
            class="bp3-button"
            type="button"
          >
            <span
              class="bp3-icon bp3-icon-chevron-up"
              icon="chevron-up"
            >
              <svg
                data-icon="chevron-up"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  chevron-up
                </desc>
                <path
                  d="M12.71 9.29l-4-4C8.53 5.11 8.28 5 8 5s-.53.11-.71.29l-4 4a1.003 1.003 0 0 0 1.42 1.42L8 7.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 0 0 .71-1.71z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
          <button
            class="bp3-button"
            type="button"
          >
            <span
              class="bp3-icon bp3-icon-chevron-down"
              icon="chevron-down"
            >
              <svg
                data-icon="chevron-down"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  chevron-down
                </desc>
                <path
                  d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 0 0-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0 0 12 5z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
